File: NtgeCore-Android/src/main/java/com/dimension/ntge/message/Decryptor.kt
package com.dimension.ntge.message

import com.dimension.ntge.Ntge
import com.dimension.ntge.x25519.X25519FileKey
import com.dimension.ntge.x25519.X25519PrivateKey

class Decryptor internal constructor(
        internal val ptr: Long
) : AutoCloseable {

    companion object {
        fun new(message: Message): Decryptor {
            return Decryptor(Ntge.newMessageDecryptor(message.ptr))
        }
    }

    fun verifyMessageMac(fileKey: X25519FileKey): Boolean {
        return Ntge.messageDecryptorVerifyMessageMac(ptr, fileKey.ptr)
    }

    fun getFileKey(privateKey: X25519PrivateKey): X25519FileKey {
        return X25519FileKey(Ntge.messageDecryptorDecryptFileKey(ptr, privateKey.ptr))
    }

    fun decryptPayload(fileKey: X25519FileKey): String {
        return Ntge.messageDecryptorDecryptPayload(ptr, fileKey.ptr).toString(Charsets.UTF_8)
    }

    fun decryptPayloadExtra(fileKey: X25519FileKey): String {
        return Ntge.decryptMessageExtra(ptr, fileKey.ptr).toString(Charsets.UTF_8)
    }

    override fun close() {
        Ntge.destroyMessageDecryptor(ptr)
    }
}

File: NtgeCore.Net/src/Message/Decryptor.cs
using System;
using NtgeCore.Net.X25519;

namespace NtgeCore.Net.Message
{
    public class Decryptor : RustObject
    {
        internal Decryptor(IntPtr ptr) : base(ptr)
        {
        }

        public static Decryptor New(NtgeMessage message)
        {
            return new Decryptor(Native.newMessageDecryptor(message.Ptr));
        }

        public bool VerifyMessageMac(X25519FileKey fileKey)
        {
            return Native.messageDecryptorVerifyMessageMac(Ptr, fileKey.Ptr);
        }

        public X25519FileKey GetFileKey(X25519PrivateKey privateKey)
        {
            return new X25519FileKey(Native.messageDecryptorDecryptFileKey(Ptr, privateKey.Ptr));
        }

        public string DecryptPayload(X25519FileKey fileKey)
        {
            using var result = Native.messageDecryptorDecryptPayload(Ptr, fileKey.Ptr);
            return result.AsString();
        }

        public string DecryptPayloadExtra(X25519FileKey fileKey)
        {
            using var result = Native.decryptMessageExtra(Ptr, fileKey.Ptr);
            return result.AsString();
        }

        public override void Dispose()
        {
            Native.destroyMessageDecryptor(Ptr);
        }
    }
}